### PR TITLE
Implement logit lens

### PR DESCRIPTION
**Pull request by llinauer**
Implement the "original" and a Patschcope logit lens.

The original logit lens too uses a Patschcope object to get access to some useful method, which I did not want to duplicate.
This is arguably not good practice and we can decouple them later on.

### CODE REVIEW

File: scripts/reproduce_logitlens_results.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
+"""
```
Comment (FergusFettes) [inline scripts/reproduce_logitlens_results.py:1]: very nice

File: obvspython/lenses.py [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+
+            # empty CDUA cache to avoid filling of GPU memory
+            torch.cuda.empty_cache()
+        self.data['substring_tokens'] = substring_tokens
+        self.data['layers'] = layers
+
+
+class ClassicLogitLens(BaseLogitLens):
+    """ Implementation of LogitLens in standard fashion.
+        Run a forward pass on the model and multiply the output of a specific layer
```
Comment (FergusFettes) [inline obvspython/lenses.py:148]: multiply here is wrong isnt it?

Comment (FergusFettes) [reply obvspython/lenses.py:148]: or you just mean to apply the final layer norm?

File: obvspython/vis.py [hunk 3]
```diff
@@ -33,3 +34,41 @@ def create_heatmap(model_1_layers, model_2_layers, values) -> go.Figure:
     )
 
     return fig
+
+
+def create_annotated_heatmap(values: List[float], cell_annotations: List[str], x_ticks: List[str],
```
Comment (FergusFettes) [inline obvspython/vis.py:39]: lovely. did you try making the other standard logit lens visses?

File: obvspython/lenses.py [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+            # loop over all layers
+            for i, layer in enumerate(self.data['layers']):
+                layer_logits = []
+                layer_preds = []
+
+                # loop over every token in substring
+                for j in range(len(self.data['substring_tokens'])):
+
+                    # get the top prediction and logit
+                    top_logit, top_pred_idx = torch.max(self.data['logits'][(i, j)], dim=0)
```
Comment (FergusFettes) [inline obvspython/lenses.py:75]: it doesnt matter much in this context, but two for loops for manipulating data in a torch tensor is kinda terrible. like, you could get the logits and preds from the tensor in one line, and then split them up.

basically the inner loop at least here should be replaced with tensor manipulation logic. looping over tensors and appending thier values to lists is just an extremely bad habit to get into when you dont have to

**Review by FergusFettes — COMMENTED** [review]
LGTM!

File: obvspython/vis.py (hunk 3 above)
Comment (llinauer) [reply obvspython/vis.py]: Not yet, but would love to in the coming sprints.

File: obvspython/lenses.py (hunk 2 above)
Comment (llinauer) [reply obvspython/lenses.py]: Yeah, I mean just apply the final layer norm. I can change it.

File: obvspython/lenses.py (hunk 4 above)
Comment (llinauer) [reply obvspython/lenses.py]: I agree. The run method of the Patchscope logit-lens should definitely be improved.
I can do so in the next sprint.

Comment (llinauer) [reply obvspython/lenses.py]: Aha, I mistook the code fragment. It's the visualization that's bad.
I will think about a better data structure here.
